share-coordinator/src/main/java/org/apache/kafka/coordinator/share/PersisterStateBatchCombiner.java

```diff
@@ -0,0 +1,394 @@
+     */
+    public List<PersisterStateBatch> combineStateBatches() {
+        pruneBatches();
+        mergeBatches();
+        return finalBatchList;
+    }
+
+    private void mergeBatches() {
+        if (combinedBatchList.size() < 2) {
+            finalBatchList = combinedBatchList;
+            return;
+        }
+
+        sortedBatches = new TreeSet<>(combinedBatchList);
+
+        BatchOverlapState overlapState = getOverlappingState();
+
+        while (overlapState != BatchOverlapState.EMPTY) {
+            PersisterStateBatch prev = overlapState.prev();
+            PersisterStateBatch candidate = overlapState.candidate();
+
+            // remove both previous and candidate for easier
+            // assessment about adding batches to sortedBatches
+            sortedBatches.remove(prev);
+            sortedBatches.remove(candidate);
+
+            if (compareBatchState(candidate, prev) == 0) {  // same state and overlap or contiguous
+                // overlap and same state (prev.firstOffset <= candidate.firstOffset) due to sort
+                // covers:
+                // case:        1        2          3            4          5           6          7 (contiguous)
+                // prev:        ------   -------    -------      -------   -------   --------    -------
+                // candidate:   ------   ----       ----------     ---        ----       -------        -------
+                handleSameStateOverlap(prev, candidate);
+            } else { // diff state and non-contiguous overlap
+                // overlap and diff state
+                // covers:
+                // case:        1        2*          3            4          5           6             7*
+                // prev:        ------   -------    -------      -------    -------     --------      ------
+                // candidate:   ------   ----       ---------      ----        ----        -------   -------
+                // max batches: 1           2       2                3          2            2          2
+                // min batches: 1           1       1                1          1            2          1
+                // * not possible with treeset
+                handleDiffStateOverlap(prev, candidate);
+            }
+            overlapState = getOverlappingState();
+        }
+        finalBatchList.addAll(sortedBatches);   // some non overlapping batches might have remained
+    }
+
+    /**
+     * Compares the non-offset state of 2 batches i.e. the deliveryCount and deliverState.
+     * <p>
+     * Uses standard compareTo contract x < y => +int, x > y => -int, x == y => 0
+     *
+     * @param b1 - {@link PersisterStateBatch} to compare
+     * @param b2 - {@link PersisterStateBatch} to compare
+     * @return int representing comparison result.
+     */
+    private int compareBatchState(PersisterStateBatch b1, PersisterStateBatch b2) {
```

Comment:
Should we just use `PersisterStateBatch.compareTo` ?

Reply:
No we explicitly want to compare the 2 state parameters only.
We decide whether the pair can be combined based on offsets but when actually merging we can 1, 2 or 3 resultant batches. This is purely determined by the delivery count and state.